src/fw/drivers/imu/lsm6dso/lsm6dso.c

```diff
@@ -16,56 +16,370 @@
+  }
+
+  if (s_lsm6dso_state.num_samples || s_lsm6dso_state.shake_detection_enabled ||
+      s_lsm6dso_state.double_tap_detection_enabled) {
+    exti_enable(BOARD_CONFIG_ACCEL.accel_ints[0]);
+  } else {
+    exti_disable(BOARD_CONFIG_ACCEL.accel_ints[0]);
+  }
+}
+
+// ODR rates helpers
+
+static odr_xl_interval_t prv_get_odr_for_interval(uint32_t interval_us) {
+  if (interval_us >= 625000) return (odr_xl_interval_t){LSM6DSO_XL_ODR_1Hz6, 625000};
+  if (interval_us >= 80000) return (odr_xl_interval_t){LSM6DSO_XL_ODR_12Hz5, 80000};
+  if (interval_us >= 38462) return (odr_xl_interval_t){LSM6DSO_XL_ODR_26Hz, 38462};
+  if (interval_us >= 19231) return (odr_xl_interval_t){LSM6DSO_XL_ODR_52Hz, 19231};
+  if (interval_us >= 9615) return (odr_xl_interval_t){LSM6DSO_XL_ODR_104Hz, 9615};
+  if (interval_us >= 4808) return (odr_xl_interval_t){LSM6DSO_XL_ODR_208Hz, 4808};
+  if (interval_us >= 2398) return (odr_xl_interval_t){LSM6DSO_XL_ODR_417Hz, 2398};
+  if (interval_us >= 1200) return (odr_xl_interval_t){LSM6DSO_XL_ODR_833Hz, 1200};
+  if (interval_us >= 600) return (odr_xl_interval_t){LSM6DSO_XL_ODR_1667Hz, 600};
+  if (interval_us >= 300) return (odr_xl_interval_t){LSM6DSO_XL_ODR_3333Hz, 300};
+  return (odr_xl_interval_t){LSM6DSO_XL_ODR_6667Hz, 150};
+}
+
+int32_t prv_lsm6dso_set_sampling_interval(uint32_t interval_us) {
+  if (!s_lsm6dso_state.initialized) {
+    PBL_LOG(LOG_LEVEL_ERROR, "LSM6DSO: Not initialized, cannot set sampling interval");
+    return -1;
+  }
+
+  odr_xl_interval_t odr_interval = prv_get_odr_for_interval(interval_us);
+  lsm6dso_xl_data_rate_set(&lsm6dso_ctx, odr_interval.odr);
+  s_lsm6dso_state.sampling_interval_us = odr_interval.interval_us;
+
+  PBL_LOG(LOG_LEVEL_DEBUG, "LSM6DSO: Set sampling interval to %lu us (requested %lu us)",
+          s_lsm6dso_state.sampling_interval_us, interval_us);
+  return odr_interval.interval_us;
+}
+
+// Read accelerometer samples from the sensor and dispatch to callback
+
+static void prv_lsm6dso_read_samples(void) {
+  // TODO: Properly implement FIFO buffer.
+  AccelDriverSample sample;
+  prv_lsm6dso_read_sample(&sample);
+}
+
+static uint8_t prv_lsm6dso_read_sample(AccelDriverSample *data) {
+  if (!s_lsm6dso_state.initialized) {
+    return -1;  // Not initialized
+    PBL_LOG(LOG_LEVEL_ERROR, "LSM6DSO: Not initialized, cannot read sample");
+  }
+
+  int16_t accel_raw[3];
+  if (lsm6dso_acceleration_raw_get(&lsm6dso_ctx, accel_raw) != 0) {
+    PBL_LOG(LOG_LEVEL_ERROR, "LSM6DSO: Failed to read accelerometer data");
+    return -1;  // Read error
+  }
+
+  data->x = lsm6dso_from_fs2_to_mg(accel_raw[0]);
+  data->y = lsm6dso_from_fs2_to_mg(accel_raw[1]);
+  data->z = lsm6dso_from_fs2_to_mg(accel_raw[2]);
+  data->timestamp_us = prv_get_timestamp_ms() * 1000;
+
+  if (s_lsm6dso_state.num_samples > 0) {
+    accel_cb_new_sample(data);
+  }
+
+  return 0;
+}
+
+// Other utility functions
+
+static void prv_delay_ms(uint32_t ms) { delay_us(ms * 1000); }
```

Comment:
`psleep`?

Reply:
I think I had some trouble with psleep for some reason. Will try again.